File: 250926/두 방향 탈출 가능 여부 판별하기/determine-escapableness-with-2-ways.java

import java.util.*;
import java.io.*;

public class Main {
	
	static int[][] visited, grid;
	static int result = 0;
	static int n, m;
	
	public static void main(String[] args) throws IOException {
		// System.setIn(new FileInputStream("src/coding_test/sample_input.txt"));
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
		StringTokenizer st = new StringTokenizer(br.readLine().trim(), " ");
		n = Integer.parseInt(st.nextToken()); // int[n][m]
		m = Integer.parseInt(st.nextToken());
		visited = new int[n][m];
		grid = new int[n][m]; // 뱀이 없을 때!! 1, 있으면 0
		
		for(int i = 0; i < n; i ++) {
			st = new StringTokenizer(br.readLine().trim(), " ");
			for(int j = 0; j < m; j++) {
				grid[i][j] = Integer.parseInt(st.nextToken());
			}
		}
		
		visited[0][0] = 1; // 시작 칸 visited 표시
		DFS(0,0);
		
		
		bw.write(String.valueOf(result));
		bw.close();
	}
	
	public static void DFS(int x, int y) {
		int[] dx = {1, 0}; //아래 오른쪽 순서로 이동가능
		int[] dy = {0, 1};
		
		for(int i = 0 ; i <2; i ++) {
			int newX = x + dx[i];
			int newY = y + dy[i];
			
			if(canGo(newX, newY)) {
				if(newX==(n-1) && newY==(m-1)) {
					result = 1;
					return; // 맨 오른쪽 아래에 canGo 상태가 판별되면 탈출가능 확인완료
				}
				visited[newX][newY] = 1;
				DFS(newX, newY);
			}
		}
	}
	
	public static boolean inRange(int x, int y) {
		return x>=0 && x<n && y>=0 && y<m;
	}
	
	public static boolean canGo(int x, int y) {
		if( !inRange(x, y)) {
			return false;
		}
		// range 안넘어가고, 뱀 없고, 방문 안했으면 갈 수 있음
		if( grid[x][y]==1 && visited[x][y]==0 ) {
			return true;
		}
		return false;
	}
}

File: 251016/두 방향 탈출 가능 여부 판별하기/determine-escapableness-with-2-ways.cpp
#include <iostream>
#include <vector>
using namespace std;

int n,m;
vector<vector<int>> grid;
vector<vector<bool>> visited;
vector<int> dx = {1, 0}; // 아래 오른쪽 위쪽 왼쪽
vector<int> dy = {0, 1};
int is_success;

bool inRange(int x, int y){
    return x>=0 && x<n && y>=0 && y<m;
}

void dfs(int x, int y){
    if(x==n-1 && y==m-1){
        is_success = 1;
        return;
    }
    int nx, ny;
    for(int i = 0; i < 2; i ++){
        nx = x+dx[i];
        ny = y+dy[i];
        // inrange, 뱀이 없고, 방문한 적 없는 칸만 dfs
        if(inRange(nx, ny) && grid[nx][ny]==1 && !visited[nx][ny]){
            visited[nx][ny] = true;
            dfs(nx, ny);
        }
    }
}

int main() {
    // ios::sync_with_stdio(false);

    cin >> n >> m;
    grid.assign(n, vector<int>(m,0)); // 초기화
    visited.assign(n, vector<bool>(m,false));

    for(int i = 0; i < n; i ++)
        for(int j = 0; j < m; j++)
            cin >> grid[i][j];
    
    is_success = 0;
    dfs(0,0);

    cout << is_success;
    return 0;
}